src/backend/egl/context.rs

```diff
@@ -385,116 +93,125 @@ impl<B: native::Backend, N: native::NativeDisplay<B>> EGLContext<B, N> {
+    /// on multiple threads.
+    pub unsafe fn make_current(&self) -> ::std::result::Result<(), SwapBuffersError> {
+        if let Some(display) = self.display.upgrade() {
+            let surface_ptr = ptr::null();
+
+            let ret = ffi::egl::MakeCurrent(
+                (*display) as *const _,
+                surface_ptr as *const _,
+                surface_ptr as *const _,
+                (*self.context) as *const _,
+            );
+
+            if ret == 0 {
+                match ffi::egl::GetError() as u32 {
+                    ffi::egl::CONTEXT_LOST => Err(SwapBuffersError::ContextLost),
+                    err => panic!("eglMakeCurrent failed (eglGetError returned 0x{:x})", err),
+                }
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(SwapBuffersError::ContextLost)
+        }
     }
 
     /// Returns true if the OpenGL context is the current one in the thread.
     pub fn is_current(&self) -> bool {
         unsafe { ffi::egl::GetCurrentContext() == (*self.context) as *const _ }
     }
 
+    /// Returns the egl config for this context
+    pub fn get_config_id(&self) -> ffi::egl::types::EGLConfig {
+        self.config_id
+    }
+
     /// Returns the pixel format of the main framebuffer of the context.
     pub fn get_pixel_format(&self) -> PixelFormat {
         self.pixel_format
     }
-
-    /// Borrow the underlying native display.
-    ///
-    /// This follows the same semantics as [`std::cell:RefCell`](std::cell::RefCell).
-    /// Multiple read-only borrows are possible. Borrowing the
-    /// backend while there is a mutable reference will panic.
-    pub fn borrow(&self) -> Ref<'_, N> {
-        self.native.borrow()
-    }
-
-    /// Borrow the underlying native display mutably.
-    ///
-    /// This follows the same semantics as [`std::cell:RefCell`](std::cell::RefCell).
-    /// Holding any other borrow while trying to borrow the backend
-    /// mutably will panic. Note that EGL will borrow the display
-    /// mutably during surface creation.
-    pub fn borrow_mut(&self) -> RefMut<'_, N> {
-        self.native.borrow_mut()
-    }
 }
 
-unsafe impl<B: native::Backend, N: native::NativeDisplay<B> + Send> Send for EGLContext<B, N> {}
-unsafe impl<B: native::Backend, N: native::NativeDisplay<B> + Sync> Sync for EGLContext<B, N> {}
-
-impl<B: native::Backend, N: native::NativeDisplay<B>> Drop for EGLContext<B, N> {
+impl Drop for EGLContext {
     fn drop(&mut self) {
         unsafe {
             // we don't call MakeCurrent(0, 0) because we are not sure that the context
             // is still the current one
-            ffi::egl::DestroyContext((*self.display) as *const _, (*self.context) as *const _);
-            ffi::egl::Terminate((*self.display) as *const _);
+            if let Some(display) = self.display.upgrade() {
+                ffi::egl::DestroyContext((*display) as *const _, (*self.context) as *const _);
```

Comment:
Something is fishy here:
- If we cannot cleanup the context without the display, the display should be no `Weak`-ptr.
- If the context is automatically destroyed, when the Display is, this is a dangling ptr. Wouldn't make_current fail in that case?

Whatever it is, I think I would prefer `EGLContext` to have a strong reference to `ffi:EGLDisplay`. Which means we need another slim wrapper around `ffi:EGLDisplay`, that is not user exposed and just implements `Deref` and `Drop` to probably destroy the display.

Same goes for the other types like `EGLSurface` or `EGLImages`.

Reply:
I am noticing right now, that this is equivalent to the old behavior.
And I also notice, that I was not smart enough to make a comment on `EGLContext`s `Drop`-implementation on how destructing in EGL works.

But while we are at it, we may as well fix that. So the open question remains.

Reply:
The `ffi::EGLDisplay` is kinda like a memory allocation group. Such that, when the display is destroyed, all resources associated with it are automatically cleaned up.

Reply:
Well then we need to use strong-pointers and a drop-wrapper. Good you are refactoring this. Could have been a long time until I would have noticed this otherwise.